### DISCUSSION TITLE
Richer log outputs to be able to correlate otel traces and logs

### DISCUSSION THREAD

**Discussion by mastoj** [vercel/next.js#71910] [Ideas]
### Goals

1. Ease production debugging
2. More insight into system performance

### Background

Open telemetry consists of three core pillars, metrics, logs and traces. I understand that metrics might not relevant from a nextjs perspective since that is more a core infrastructure part. When it comes to logs and traces it should be an easy way to enrich what is logged and structure for easier correlation between logs and traces. What I would expect to see in logs when I have instrumentation on is my logs to contain span_id and trace_id, which would enable a lot of the otel providers out there to correlate the logs with my traces.

### Proposal

I'm not 100% sure if this is really nextjs or if it is vercel specific to be honest. I will definitely contribute verifying, if time allows I can try to help out developing.

**Comment by aniruddhaadak80** [comment]
Hi @mastoj ,


Thank you for highlighting the need for richer log outputs in Next.js to enhance production debugging and system performance insight. Including `span_id` and `trace_id` in the logs would indeed facilitate better correlation between logs and traces when using OpenTelemetry.

To address this, I suggest the following steps:

1. **Log Enrichment**: Implement log enrichment to include `span_id` and `trace_id` in your logging configuration. This may require custom logging middleware that captures and formats these identifiers.
   
2. **OpenTelemetry Integration**: Ensure that your OpenTelemetry setup correctly propagates these identifiers through your application, which can typically be configured in the instrumentation settings.


Thanks .

**Reply by mmanciop** [reply]
Hi all,

I think the runtime logs emitted by Vercel through the log drains should include (either _by default_ or opt-in) the `trace_id` and `span_id` if the OpenTelemetry tracing in Vercel is being used (https://vercel.com/docs/observability/otel-overview). This is not something that users would likely be able to duplicate on their own, especially for error scenarios (which are the most interesting).
